### PR TITLE
Fixing issue #746

### LINKED ISSUE #746: [Bug] For a language without a Pygments lexer, the program crashes
# Expected Behavior

A language that may not have a style coloring lexer would either not use one or use some default of another language.

# Actual Behavior

CT crashes.

image

# Steps to Reproduce the Problem

1. Run PR #744 locally
2. Do a reference of CLIPS with the Rules concept
3. Crash will happen

# Additional Info

I'm not sure what the right answer is or if there's a "default" lexer ([I didn't see one](https://pygments.org/languages/) but didn't look too hard). Maybe no formatting if there's no lexer?

### PR DESCRIPTION
**Pull request by ayaan-qadri**
## What GitHub issue does this PR apply to?

 

Resolves #746

## What changed and why?

There was an issue with Pygments because it does not support 'clips' highlighting. After some research, I found that 'prolog' has a similar syntax highlighting style. Therefore, I modified the get_lexer_by_name function to check for 'clips' and changed it with 'prolog'. I tested the changes, and the issue has been resolved, with syntax highlighting working as expected.

## (If editing website code) Please add screenshots

image

127 0 0 1_8000-CodeThesaurus-ReferenceforCLIPS(641)

## Checklist

   

- [ ] I claimed any associated issue(s) and they are not someone else's
- [X] I have looked at documentation to ensure I made any revisions correctly
- [X] I tested my changes locally to ensure they work
- [ ] For language files, I have validated the edited files are valid JSON and data shows up correctly
- [ ] For website code edits, I have added or edited any appropriate unit tests for my changes

### CODE REVIEW

File: web/thesauruses/_meta/rules_facts.json [hunk 1]
```diff
@@ -0,0 +1,20 @@
+{
```
Comment (geekygirlsarah) [inline web/thesauruses/_meta/rules_facts.json]: Including this file will cause a conflict with #744, so I'd rather this PR just fix the bug, then that PR can add in the rules/facts concepts. So I'd delete this file.

File: web/thesauruses/clips/641/rules_facts.json [hunk 2]
```diff
@@ -0,0 +1,32 @@
+{
```
Comment (geekygirlsarah) [inline web/thesauruses/clips/641/rules_facts.json]: Same for this file, it comes from #744.

File: web/thesauruses/meta_info.json [hunk 3]
```diff
@@ -5,6 +5,7 @@
     "c": "C",
```
Comment (geekygirlsarah) [inline web/thesauruses/meta_info.json]: And same for these changes.

**Review by geekygirlsarah — CHANGES_REQUESTED** [review]
This is a fix I'm willing to take, but it shouldn't duplicate the effort of #744. Take out those changes and just leave `views.py` changes and I'll merge this in!

File: web/thesauruses/meta_info.json (hunk 3 above)
Comment (ayaan-qadri) [reply web/thesauruses/meta_info.json:5]: @geekygirlsarah Sure thing, I will just do necessary change which is in view.py asap, thank you !

**Comment by ayaan-qadri** [comment]
@geekygirlsarah Sure thing, I will just do necessary change which is in view.py asap, thank you !

**Comment by ayaan-qadri** [comment]
Here’s a grammatically corrected version of your message:

Hey @geekygirlsarah , I'm currently checking for the 'clips' language only. Would you like to make Prolog the default for syntax highlighting?

Advantages:
If we set Prolog as the default language, then when we add any new languages in the future that aren't supported by Pygments, this same error won't occur again.

Disadvantages:
Syntax highlighting for new languages can't be guaranteed. It will only work if the new language's syntax is similar to Prolog's.

**Comment by ayaan-qadri** [comment]
Hey @geekygirlsarah , I'm currently checking for the 'clips' language only. Would you like to make Prolog the default for syntax highlighting?

Advantages:
If we set Prolog as the default language, then when we add any new languages in the future that aren't supported by Pygments, this same error won't occur again.

Disadvantages:
Syntax highlighting for new languages can't be guaranteed. It will only work if the new language's syntax is similar to Prolog's.

**Comment by geekygirlsarah** [comment]
> Would you like to make Prolog the default for syntax highlighting?

I'd say no. Maybe just the plain text lexer instead?

image

**Comment by ayaan-qadri** [comment]
> > Would you like to make Prolog the default for syntax highlighting?
> 
> I'd say no. Maybe just the plain text lexer instead?
> 
> image

Sure, then I will set plain text lexer as default if it's not exists.

File: web/views.py [hunk 4]
```diff
@@ -307,6 +309,15 @@ def error_handler_500_server_error(request):
     response = render(request, 'error500.html')
     return HttpResponseServerError(response)
 
+#get lexer 
+def get_highlighter(language):
+    SIMILAR_LEXERS = settings.SIMILAR_LEXERS
+    try:
+        lexer = get_lexer_by_name(language, startinline=True)
+    except ClassNotFound:
+
```
Comment (geekygirlsarah) [inline web/views.py]: Want to remove this blank line?

**Review by geekygirlsarah — CHANGES_REQUESTED** [review]
Interesting change, I was thinking of moving the lexers to the meta file for languages to keep language stuff together, but I guess it could work in settings too. I can merge it in if you tidy up the try/except blank line!

**Comment by ayaan-qadri** [comment]
So, i created a new function named 'get_highlighter' which returns lexer and added SIMILAR_LEXERS in setings.py so now if this case happens again we just have to add into dictionary ```<language_which_is_not_support_in_pygments> : <similar_syntax_langauge_supported_in_pygments>``` if it exists in SIMILAR_LEXERS it will pick that else default it will be plain 'text' lexer.

File: web/views.py (hunk 4 above)
Comment (ayaan-qadri) [reply web/views.py]: Sorry, My bad. Done !

**Comment by ayaan-qadri** [comment]
> Interesting change, I was thinking of moving the lexers to the meta file for languages to keep language stuff together, but I guess it could work in settings too. I can merge it in if you tidy up the try/except blank line!

I chose to put it in settings because the settings are loaded into memory when the app starts, which is comparatively faster than anything else !

**Review by geekygirlsarah — APPROVED** [review]
I'll merge it in! Thanks for helping out!

Don't forget it's Hacktoberfest this month. Be sure to register if you haven't as this project counts towards it!

**Pull request merged**